x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SwapAliasesAndDeleteSourceIndexStep.java

```diff
@@ -0,0 +1,91 @@
+ * source index.
+ */
+public class SwapAliasesAndDeleteSourceIndexStep extends AsyncActionStep {
+    public static final String NAME = "swap-aliases-to-restored";
+    private static final Logger logger = LogManager.getLogger(WaitForGreenIndexHealthStep.class);
+
+    private final String restoredIndexPrefix;
+
+    public SwapAliasesAndDeleteSourceIndexStep(StepKey key, StepKey nextStepKey, Client client, String restoredIndexPrefix) {
+        super(key, nextStepKey, client);
+        this.restoredIndexPrefix = restoredIndexPrefix;
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    @Override
+    public void performAction(IndexMetaData indexMetaData, ClusterState currentClusterState, ClusterStateObserver observer,
+                              Listener listener) {
+        String originalIndex = indexMetaData.getIndex().getName();
+        final String restoredIndexName = restoredIndexPrefix + originalIndex;
+        IndexMetaData restoredIndexMetaData = currentClusterState.metaData().index(restoredIndexName);
+
+        if (restoredIndexMetaData == null) {
+            String policyName = indexMetaData.getSettings().get(LifecycleSettings.LIFECYCLE_NAME);
+            String errorMessage = String.format(Locale.ROOT, "restored index [%s] doesn't exist. stopping execution of lifecycle [%s] for" +
+                " index [%s]", restoredIndexName, policyName, originalIndex);
+            logger.debug(errorMessage);
+            listener.onFailure(new IllegalStateException(errorMessage));
+            return;
+        }
+
+        // TODO extract and reuse - also see {@link org.elasticsearch.xpack.core.ilm.ShrinkSetAliasStep}
+        IndicesAliasesRequest aliasesRequest = new IndicesAliasesRequest()
+            .masterNodeTimeout(getMasterTimeout(currentClusterState))
+            .addAliasAction(IndicesAliasesRequest.AliasActions.removeIndex().index(originalIndex))
+            .addAliasAction(IndicesAliasesRequest.AliasActions.add().index(restoredIndexName).alias(originalIndex));
+        indexMetaData.getAliases().values().spliterator().forEachRemaining(aliasMetaDataObjectCursor -> {
+            AliasMetaData aliasMetaDataToAdd = aliasMetaDataObjectCursor.value;
+            // inherit all alias properties except `is_write_index`
+            aliasesRequest.addAliasAction(IndicesAliasesRequest.AliasActions.add()
+                .index(restoredIndexName).alias(aliasMetaDataToAdd.alias())
+                .indexRouting(aliasMetaDataToAdd.indexRouting())
+                .searchRouting(aliasMetaDataToAdd.searchRouting())
+                .filter(aliasMetaDataToAdd.filter() == null ? null : aliasMetaDataToAdd.filter().string())
+                .writeIndex(aliasMetaDataToAdd.writeIndex()));
+        });
+
+        getClient().admin().indices().aliases(aliasesRequest, new ActionListener<>() {
```

Comment:
If you want, in the future, you can use `ActionListener.wrap(resp -> listener.onResponse(true), listener::onFailure)` I believe, rather than a new anonymous class